core/src/main/scala/com/salesforce/op/stages/impl/selector/RandomParamBuilder.scala

```diff
@@ -0,0 +1,195 @@
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.salesforce.op.stages.impl.selector
+
+import com.salesforce.op.stages.impl.selector.Distribution.{Exponential, Subset, Uniform}
+import enumeratum._
+import org.apache.spark.ml.param._
+
+import scala.util.Random
+import scala.collection.mutable
+
+sealed abstract class Distribution extends EnumEntry with Serializable
+
+private object Distribution extends Enum[Distribution] {
+  val values: Seq[Distribution] = findValues
+  case object Uniform extends Distribution
+  case object Subset extends Distribution
+  case object Exponential extends Distribution // TODO inverse exponential?
+}
+
+/**
+ * Builder for a param sets used in random search-based model selection.
+ */
+class RandomParamBuilder {
+
+  private val paramDefs = mutable.Map.empty[Param[_], (Distribution, _, _, Seq[_])]
+  private val rand = new Random()
+
+  private def addParams[T](param: Param[T], dist: Distribution, min: T, max: T): this.type = {
+    paramDefs.put(param, (dist, min, max, Seq()))
+    this
+  }
+
+
+  /**
+   * Adds parameter values uniformly selected from a sequence
+   *
+   * @param param parameter to add
+   *
+   * @param seq   sequence of possible values
+   */
+  def subset[T](param: Param[T], seq: Seq[T]): this.type = {
+    paramDefs.put(param, (Subset, None, None, seq))
+    this
+  }
+
+  /**
+   * Adds double param with range of values in uniform distribution
+   *
+   * @param param parameter to add
+   * @param min   minimum value for param
+   * @param max   maximum value for param
+   */
+  def uniform(param: DoubleParam, min: Double, max: Double): this.type = {
+    assert(min < max, "min must be less than max")
```

Comment:
use `require` instead of `assert`, as assertions can be skipped at runtime and requires cannot.